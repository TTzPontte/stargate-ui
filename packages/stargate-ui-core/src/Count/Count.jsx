import React from 'react';
import PropTypes from 'prop-types';
import { useCountUp } from 'react-countup';

import Factory from '../Factory';

const Count = React.forwardRef((props, ref) => {
  const {
    onChange,
    delay = 0,
    duration = 5,
    decimals = 2,
    children = '',
    style = 'decimal',
    currency = 'BRL',
    locale: numberFormatLocale = 'pt-BR',
    ...factoryProps
  } = props;

  const number = children
      /**
       * Remove non-numeric and non-comma character.
       */
      .replace(/[^\d\,]/g, '')
      /**
       * Replace comma for dot.
       */
      .replace(/\,/, '.');
  const end = parseFloat(number);

  const handleFormat = (n) => {
    if (onChange) {
      onChange({}, n);
    }

    const numberFormatOptions = { style, minimumFractionDigits: decimals };

    if (style === 'currency' && currency) {
      numberFormatOptions.currency = currency;
    }
    /**
     * @todo create format component or utils.
     */
    const valueFormatted = new Intl.NumberFormat(numberFormatLocale, numberFormatOptions).format(n);

    return valueFormatted;
  }

  const { countUp } = useCountUp({
    duration,
    decimals,
    end,
    delay,
    start: 0,
    formattingFn: handleFormat,
  });

  return (
    <Factory element="span" {...factoryProps} ref={ref}>
      {countUp}
    </Factory>
  );
});

Count.displayName = 'Count';

Count.propTypes = {
  /**
   * Add string node. It does not work DOM elements or React components.
   */
  children: PropTypes.string,
  /**
   * Add delay in seconds before start.
   * @see {@link https://github.com/glennreyes/react-countup#delay-number}
   * @default 0
   */
  delay: PropTypes.number,
  /**
   * Add animation duration in seconds.
   * @see {@link https://github.com/glennreyes/react-countup#duration-number}
   * @default 0
   */
  duration: PropTypes.number,
  /**
   * Add amount of decimals. It used by both CountUp and Intl.NumberFormat.
   * @see {@link https://github.com/glennreyes/react-countup#decimals-number}
   * @see {@link https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat}
   * @default 2
   */
  decimals: PropTypes.number,
  /**
   * Trigger when children animation is changing.
   * @default undefined
   */
  onChange: PropTypes.func,
  /**
   * Add format style. It used by Intl.NumberFormat and accepts its specified values.
   * @see {@link https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat}
   * @default 'decimal'
   */
  style: PropTypes.oneOf([
    'decimal',
    'percent',
    'currency',
  ]),
  /**
   * Add currency format. It used by Intl.NumberFormat and accepts its specified values.
   * @see {@link https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat}
   * @default 'BRL'
   */
  currency: PropTypes.string,
  /**
   * Add locale format. It used by Intl.NumberFormat and accepts its specified values.
   * @see {@link https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat}
   * @default 'pt-BR'
   */
  locale: PropTypes.string,
  /**
   * Add any props allowed in Factory or default props of HTML.
   */
  // '...props': PropTypes.object,
};

export default Count;
