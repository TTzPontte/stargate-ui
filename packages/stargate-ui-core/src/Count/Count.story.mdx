import {
  Meta,
  Canvas,
  Status,
  Props,
  Story,
} from '@pontte/stargate-ui-playground';

import Typography from '../Typography';
import Count from './Count';

<Meta title="Core/Count" component={Count} />

# Count

<Status type="stable" />

## Basic Usage

<Canvas>
  <Story name="Basic Usage">
    <Typography gutter={0}>
      <Count>123456,62</Count>
    </Typography>
  </Story>
</Canvas>

<Props component={Count} />

## Number format

Count accepts **decimal**, **currency** and **percent** format. It uses [Intl.NumberFormat](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat) constructor from JavaScript API to format number.

Count will assume pt-BR &#127463;&#127479; locale if <b>locale</b> and <b>currency</b> is not defined.

<Canvas>
  <Typography gutter={0}>
    <Count style="currency">123456,62</Count>
  </Typography>
</Canvas>

Using <b>locale</b> and <b>currency</b> for en-US &#127482;&#127480; format.

<Canvas>
  <Typography gutter={0}>
    <Count
      style="currency"
      locale="en-US"
      currency="USD"
    >
      123456,62
    </Count>
  </Typography>
</Canvas>

## Working with changes

In a case when you need to update a string when number hit a target, you can use `onChange` event to retrieve current number.

<Canvas>
  <Story name="Working with changes">
    {
      (props) => {
        const [count, setCount] = React.useState(0);
        const handleChange = (event, value) => {
          setCount(value)
        };
        return (
          <>
            <Typography element="span">
              <Count style="currency" onChange={handleChange}>
                3.000.000,00
              </Count>
              &nbsp;{count >= 2000000 ? 'milions' : 'milion'}!
            </Typography>
          </>
        );
      }
    }
  </Story>
</Canvas>
