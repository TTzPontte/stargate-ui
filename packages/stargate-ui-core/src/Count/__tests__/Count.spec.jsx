import React from 'react';

import Count from '../Count';

jest.mock('countup.js', () => ({
  CountUp: jest.fn().mockImplementation(() => ({
    start: jest.fn,
  })),
}));

describe('Count', () => {
  it('should render properly', () => {
    const { container } = render(
      <Count
        /**
         * Prevent Intl error in CI environment.
         */
        currency="USD"
        locale="en-US"
      >
        123.456,00
      </Count>,
    );

    expect(container).toMatchSnapshot();
  });
});
