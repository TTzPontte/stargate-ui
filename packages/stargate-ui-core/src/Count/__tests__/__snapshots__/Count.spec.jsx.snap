// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Count should render properly 1`] = `
<div>
  <span
    class="stargate-ui-factory-0-2-1 stargate-ui-factory-d0-0-2-3"
  >
    <span>
      0.00
    </span>
  </span>
</div>
`;
